# Data Manipulation - Good Practices

This notebook aims at showing good practices to `manipulate data` using Python's most popular libraries. The following are covered:
* cleaning data with `pandas`
* make specific changes with `numpy`
* handling date-related values with `datetime`

In [1]:
# imporintg required packages
import os
import pandas as pd
import numpy as np
import datetime as dt

# suppress warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# set option to view all columns
pd.set_option('display.max_columns', None)

## Data Preprocessing

In [3]:
# load data
flights_df_raw = pd.read_csv('flights.csv')
airports_df = pd.read_csv('airports.csv')
airlines_df = pd.read_csv('airlines.csv')

# view shape
print(flights_df_raw.shape)
print(airports_df.shape)
print(airlines_df.shape)

(5819079, 31)
(322, 7)
(14, 2)


In [4]:
# view top 5 rows
flights_df_raw.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  DEPARTURE_TIME  \
0            ANC                 SEA                    5          2354.0   
1            LAX                 PBI                   10             2.0   
2            SFO                 CLT                   20            18.0   
3            LAX                 MIA                   20            15.0   
4            SEA                 ANC                   25            24.0   

   DEPARTURE_DELAY  TAXI_OUT  WHEELS_OFF  SCHEDULED_TIME  ELAPSED_TIME  \
0            -11.0      21.0        15.0           205.0         194.0   
1             -8.0      12.0        14.0           280.0         279.0   
2             -2.0      16.0        34.0           286.0         293.0   
3             -5.0      15.0        30.0           285.0         281.0   
4             -1.0      11.0        35.0           235.0         215.0   

   AIR_TIME  DISTANCE  WHEELS_ON  TAXI_IN  SCHEDULED_ARRIVAL  ARRIVAL_TIME  \
0     169.0      1448      404.0      4.0                430         408.0   
1     263.0      2330      737.0      4.0                750         741.0   
2     266.0      2296      800.0     11.0                806         811.0   
3     258.0      2342      748.0      8.0                805         756.0   
4     199.0      1448      254.0      5.0                320         259.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          -22.0         0          0                 NaN               NaN   
1           -9.0         0          0                 NaN               NaN   
2            5.0         0          0                 NaN               NaN   
3           -9.0         0          0                 NaN               NaN   
4          -21.0         0          0                 NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             NaN            NaN                  NaN            NaN  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN

In [5]:
# convert feature names to lowercase
flights_df_raw.columns = flights_df_raw.columns.str.lower()
flights_df_raw.head()

year  month  day  day_of_week airline  flight_number tail_number  \
0  2015      1    1            4      AS             98      N407AS   
1  2015      1    1            4      AA           2336      N3KUAA   
2  2015      1    1            4      US            840      N171US   
3  2015      1    1            4      AA            258      N3HYAA   
4  2015      1    1            4      AS            135      N527AS   

  origin_airport destination_airport  scheduled_departure  departure_time  \
0            ANC                 SEA                    5          2354.0   
1            LAX                 PBI                   10             2.0   
2            SFO                 CLT                   20            18.0   
3            LAX                 MIA                   20            15.0   
4            SEA                 ANC                   25            24.0   

   departure_delay  taxi_out  wheels_off  scheduled_time  elapsed_time  \
0            -11.0      21.0        15.0           205.0         194.0   
1             -8.0      12.0        14.0           280.0         279.0   
2             -2.0      16.0        34.0           286.0         293.0   
3             -5.0      15.0        30.0           285.0         281.0   
4             -1.0      11.0        35.0           235.0         215.0   

   air_time  distance  wheels_on  taxi_in  scheduled_arrival  arrival_time  \
0     169.0      1448      404.0      4.0                430         408.0   
1     263.0      2330      737.0      4.0                750         741.0   
2     266.0      2296      800.0     11.0                806         811.0   
3     258.0      2342      748.0      8.0                805         756.0   
4     199.0      1448      254.0      5.0                320         259.0   

   arrival_delay  diverted  cancelled cancellation_reason  air_system_delay  \
0          -22.0         0          0                 NaN               NaN   
1           -9.0         0          0                 NaN               NaN   
2            5.0         0          0                 NaN               NaN   
3           -9.0         0          0                 NaN               NaN   
4          -21.0         0          0                 NaN               NaN   

   security_delay  airline_delay  late_aircraft_delay  weather_delay  
0             NaN            NaN                  NaN            NaN  
1             NaN            NaN                  NaN            NaN  
2             NaN            NaN                  NaN            NaN  
3             NaN            NaN                  NaN            NaN  
4             NaN            NaN                  NaN            NaN

In [6]:
# view top 5 rows
airports_df.columns = airports_df.columns.str.lower()
airports_df.head()

iata_code                              airport         city state country  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   latitude  longitude  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [7]:
# view top 5 rows
airlines_df.columns = airlines_df.columns.str.lower()
airlines_df.head()

iata_code                 airline
0        UA   United Air Lines Inc.
1        AA  American Airlines Inc.
2        US         US Airways Inc.
3        F9  Frontier Airlines Inc.
4        B6         JetBlue Airways

In [8]:
# view data types and first rows of data frame
flights_df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819079 entries, 0 to 5819078
Data columns (total 31 columns):
 #   Column               Dtype  
---  ------               -----  
 0   year                 int64  
 1   month                int64  
 2   day                  int64  
 3   day_of_week          int64  
 4   airline              object 
 5   flight_number        int64  
 6   tail_number          object 
 7   origin_airport       object 
 8   destination_airport  object 
 9   scheduled_departure  int64  
 10  departure_time       float64
 11  departure_delay      float64
 12  taxi_out             float64
 13  wheels_off           float64
 14  scheduled_time       float64
 15  elapsed_time         float64
 16  air_time             float64
 17  distance             int64  
 18  wheels_on            float64
 19  taxi_in              float64
 20  scheduled_arrival    int64  
 21  arrival_time         float64
 22  arrival_delay        float64
 23  diverted             int64  
 24

From the output above, we see that the data frame has `5819079` observations and `9` dimensions that inform on when the flight happened, the origin and destination, any potential delays as well as other miscellaneous indicators.

## Check for Data Quality

We want to see if there are any outliers, absurd values or missing values.

In [9]:
# compute statistics of columns
flights_df_raw.describe()

year         month           day   day_of_week  flight_number  \
count  5819079.0  5.819079e+06  5.819079e+06  5.819079e+06   5.819079e+06   
mean      2015.0  6.524085e+00  1.570459e+01  3.926941e+00   2.173093e+03   
std          0.0  3.405137e+00  8.783425e+00  1.988845e+00   1.757064e+03   
min       2015.0  1.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00   
25%       2015.0  4.000000e+00  8.000000e+00  2.000000e+00   7.300000e+02   
50%       2015.0  7.000000e+00  1.600000e+01  4.000000e+00   1.690000e+03   
75%       2015.0  9.000000e+00  2.300000e+01  6.000000e+00   3.230000e+03   
max       2015.0  1.200000e+01  3.100000e+01  7.000000e+00   9.855000e+03   

       scheduled_departure  departure_time  departure_delay      taxi_out  \
count         5.819079e+06    5.732926e+06     5.732926e+06  5.730032e+06   
mean          1.329602e+03    1.335204e+03     9.370158e+00  1.607166e+01   
std           4.837518e+02    4.964233e+02     3.708094e+01  8.895574e+00   
min           1.000000e+00    1.000000e+00    -8.200000e+01  1.000000e+00   
25%           9.170000e+02    9.210000e+02    -5.000000e+00  1.100000e+01   
50%           1.325000e+03    1.330000e+03    -2.000000e+00  1.400000e+01   
75%           1.730000e+03    1.740000e+03     7.000000e+00  1.900000e+01   
max           2.359000e+03    2.400000e+03     1.988000e+03  2.250000e+02   

         wheels_off  scheduled_time  elapsed_time      air_time      distance  \
count  5.730032e+06    5.819073e+06  5.714008e+06  5.714008e+06  5.819079e+06   
mean   1.357171e+03    1.416859e+02  1.370062e+02  1.135116e+02  8.223565e+02   
std    4.980094e+02    7.521058e+01  7.421107e+01  7.223082e+01  6.077843e+02   
min    1.000000e+00    1.800000e+01  1.400000e+01  7.000000e+00  2.100000e+01   
25%    9.350000e+02    8.500000e+01  8.200000e+01  6.000000e+01  3.730000e+02   
50%    1.343000e+03    1.230000e+02  1.180000e+02  9.400000e+01  6.470000e+02   
75%    1.754000e+03    1.730000e+02  1.680000e+02  1.440000e+02  1.062000e+03   
max    2.400000e+03    7.180000e+02  7.660000e+02  6.900000e+02  4.983000e+03   

          wheels_on       taxi_in  scheduled_arrival  arrival_time  \
count  5.726566e+06  5.726566e+06       5.819079e+06  5.726566e+06   
mean   1.471469e+03  7.434971e+00       1.493808e+03  1.476491e+03   
std    5.221879e+02  5.638548e+00       5.071647e+02  5.263197e+02   
min    1.000000e+00  1.000000e+00       1.000000e+00  1.000000e+00   
25%    1.054000e+03  4.000000e+00       1.110000e+03  1.059000e+03   
50%    1.509000e+03  6.000000e+00       1.520000e+03  1.512000e+03   
75%    1.911000e+03  9.000000e+00       1.918000e+03  1.917000e+03   
max    2.400000e+03  2.480000e+02       2.400000e+03  2.400000e+03   

       arrival_delay      diverted     cancelled  air_system_delay  \
count   5.714008e+06  5.819079e+06  5.819079e+06      1.063439e+06   
mean    4.407057e+00  2.609863e-03  1.544643e-02      1.348057e+01   
std     3.927130e+01  5.102012e-02  1.233201e-01      2.800368e+01   
min    -8.700000e+01  0.000000e+00  0.000000e+00      0.000000e+00   
25%    -1.300000e+01  0.000000e+00  0.000000e+00      0.000000e+00   
50%    -5.000000e+00  0.000000e+00  0.000000e+00      2.000000e+00   
75%     8.000000e+00  0.000000e+00  0.000000e+00      1.800000e+01   
max     1.971000e+03  1.000000e+00  1.000000e+00      1.134000e+03   

       security_delay  airline_delay  late_aircraft_delay  weather_delay  
count    1.063439e+06   1.063439e+06         1.063439e+06   1.063439e+06  
mean     7.615387e-02   1.896955e+01         2.347284e+01   2.915290e+00  
std      2.143460e+00   4.816164e+01         4.319702e+01   2.043334e+01  
min      0.000000e+00   0.000000e+00         0.000000e+00   0.000000e+00  
25%      0.000000e+00   0.000000e+00         0.000000e+00   0.000000e+00  
50%      0.000000e+00   2.000000e+00         3.000000e+00   0.000000e+00  
75%      0.000000e+00   1.900000e+01         2.900000e+01   0.000000e+00  
max      5.730000e+02   1.971000

In [10]:
# check for missing values
flights_df_raw.isnull().sum().sort_values(ascending=False)

cancellation_reason    5729195
weather_delay          4755640
late_aircraft_delay    4755640
airline_delay          4755640
security_delay         4755640
air_system_delay       4755640
air_time                105071
arrival_delay           105071
elapsed_time            105071
wheels_on                92513
taxi_in                  92513
arrival_time             92513
taxi_out                 89047
wheels_off               89047
departure_delay          86153
departure_time           86153
tail_number              14721
scheduled_time               6
scheduled_departure          0
cancelled                    0
day                          0
day_of_week                  0
airline                      0
flight_number                0
scheduled_arrival            0
diverted                     0
origin_airport               0
distance                     0
destination_airport          0
month                        0
year                         0
dtype: int64

In [11]:
# check for percentage of missing values
round(flights_df_raw.isnull().sum().sort_values(ascending=False)*100/flights_df_raw.shape[0], 2)

cancellation_reason    98.46
weather_delay          81.72
late_aircraft_delay    81.72
airline_delay          81.72
security_delay         81.72
air_system_delay       81.72
air_time                1.81
arrival_delay           1.81
elapsed_time            1.81
wheels_on               1.59
taxi_in                 1.59
arrival_time            1.59
taxi_out                1.53
wheels_off              1.53
departure_delay         1.48
departure_time          1.48
tail_number             0.25
scheduled_time          0.00
scheduled_departure     0.00
cancelled               0.00
day                     0.00
day_of_week             0.00
airline                 0.00
flight_number           0.00
scheduled_arrival       0.00
diverted                0.00
origin_airport          0.00
distance                0.00
destination_airport     0.00
month                   0.00
year                    0.00
dtype: float64

In [12]:
def convert_to_timedate(x, col='scheduled_departure'):
    'Converts to timestamp'

    return (
        pd.to_datetime(
            x['year'].astype(str) + '-' + x['month'].astype(str) + '-' + x['day'].astype(str) +
            ' ' + x.assign(zero='0')['zero'] * (4 - x[col].astype(str).str.len()) +
            x[col].astype(str), format='%Y-%m-%d %H%M', errors='coerce'
        )
    )

In [13]:
flights_df = (
    flights_df_raw
    # All rows should not have any null value
    .dropna(how='all')
    # Only flights from set of airports and with reasonable delay amount
    [lambda x:
         (x['origin_airport'].isin(['BOS', 'SFO', 'LAX', 'JFK'])) &
         (x['departure_delay'] <= 24*60)
    ]
    # Fix columns
    .assign(
        flight_number=lambda x: x['flight_number'].astype(str),
        scheduled_departure=lambda x: convert_to_timedate(x),
        is_delayed=lambda x: np.select(
            [x['departure_delay'] >= 15], [1],
            default=0
        )
    )
    # Remove redundant columns
    .drop(['year', 'month', 'day'], axis=1)
)
flights_df.shape

(535744, 29)

In [14]:
flights_df.head()

day_of_week airline flight_number tail_number origin_airport  \
1            4      AA          2336      N3KUAA            LAX   
2            4      US           840      N171US            SFO   
3            4      AA           258      N3HYAA            LAX   
5            4      DL           806      N3730B            SFO   
7            4      US          2013      N584UW            LAX   

  destination_airport scheduled_departure  departure_time  departure_delay  \
1                 PBI 2015-01-01 00:10:00             2.0             -8.0   
2                 CLT 2015-01-01 00:20:00            18.0             -2.0   
3                 MIA 2015-01-01 00:20:00            15.0             -5.0   
5                 MSP 2015-01-01 00:25:00            20.0             -5.0   
7                 CLT 2015-01-01 00:30:00            44.0             14.0   

   taxi_out  wheels_off  scheduled_time  elapsed_time  air_time  distance  \
1      12.0        14.0           280.0         279.0     263.0      2330   
2      16.0        34.0           286.0         293.0     266.0      2296   
3      15.0        30.0           285.0         281.0     258.0      2342   
5      18.0        38.0           217.0         230.0     206.0      1589   
7      13.0        57.0           273.0         249.0     228.0      2125   

   wheels_on  taxi_in  scheduled_arrival  arrival_time  arrival_delay  \
1      737.0      4.0                750         741.0           -9.0   
2      800.0     11.0                806         811.0            5.0   
3      748.0      8.0                805         756.0           -9.0   
5      604.0      6.0                602         610.0            8.0   
7      745.0      8.0                803         753.0          -10.0   

   diverted  cancelled cancellation_reason  air_system_delay  security_delay  \
1         0          0                 NaN               NaN             NaN   
2         0          0                 NaN               NaN             NaN   
3         0          0                 NaN               NaN             NaN   
5         0          0                 NaN               NaN             NaN   
7         0          0                 NaN               NaN             NaN   

   airline_delay  late_aircraft_delay  weather_delay  is_delayed  
1            NaN                  NaN            NaN           0  
2            NaN                  NaN            NaN           0  
3            NaN                  NaN            NaN           0  
5            NaN                  NaN            NaN           0  
7            NaN                  NaN            NaN           0

## Joins, Aggregations

### Get airport name

We want to obtain the airport name corresponding to the airport code attached to flights.

In [15]:
# airport names
merged_df = flights_df.merge(airports_df, left_on='origin_airport', right_on='iata_code')
merged_df.head()

day_of_week airline flight_number tail_number origin_airport  \
0            4      AA          2336      N3KUAA            LAX   
1            4      AA           258      N3HYAA            LAX   
2            4      US          2013      N584UW            LAX   
3            4      DL          1434      N547US            LAX   
4            4      AA           115      N3CTAA            LAX   

  destination_airport scheduled_departure  departure_time  departure_delay  \
0                 PBI 2015-01-01 00:10:00             2.0             -8.0   
1                 MIA 2015-01-01 00:20:00            15.0             -5.0   
2                 CLT 2015-01-01 00:30:00            44.0             14.0   
3                 MSP 2015-01-01 00:35:00            35.0              0.0   
4                 MIA 2015-01-01 01:05:00           103.0             -2.0   

   taxi_out  wheels_off  scheduled_time  elapsed_time  air_time  distance  \
0      12.0        14.0           280.0         279.0     263.0      2330   
1      15.0        30.0           285.0         281.0     258.0      2342   
2      13.0        57.0           273.0         249.0     228.0      2125   
3      18.0        53.0           214.0         210.0     188.0      1535   
4      14.0       117.0           286.0         276.0     255.0      2342   

   wheels_on  taxi_in  scheduled_arrival  arrival_time  arrival_delay  \
0      737.0      4.0                750         741.0           -9.0   
1      748.0      8.0                805         756.0           -9.0   
2      745.0      8.0                803         753.0          -10.0   
3      601.0      4.0                609         605.0           -4.0   
4      832.0      7.0                851         839.0          -12.0   

   diverted  cancelled cancellation_reason  air_system_delay  security_delay  \
0         0          0                 NaN               NaN             NaN   
1         0          0                 NaN               NaN             NaN   
2         0          0                 NaN               NaN             NaN   
3         0          0                 NaN               NaN             NaN   
4         0          0                 NaN               NaN             NaN   

   airline_delay  late_aircraft_delay  weather_delay  is_delayed iata_code  \
0            NaN                  NaN            NaN           0       LAX   
1            NaN                  NaN            NaN           0       LAX   
2            NaN                  NaN            NaN           0       LAX   
3            NaN                  NaN            NaN           0       LAX   
4            NaN                  NaN            NaN           0       LAX   

                             airport         city state country  latitude  \
0  Los Angeles International Airport  Los Angeles    CA     USA  33.94254   
1  Los Angeles International Airport  Los Angeles    CA     USA  33.94254   
2  Los Angeles International Airport  Los Angeles    CA     USA  33.94254   
3  Los Angeles International Airport  Los Angeles    CA     USA  33.94254   
4  Los Angeles International Airport  Los Angeles    CA     USA  33.94254   

   longitude  
0 -118.40807  
1 -118.40807  
2 -118.40807  
3 -118.40807  
4 -118.40807

### Number of flights per airport

We want to obtain the number of departing flights per airport across the year.

In [16]:
# number of departing flights per airport across the year
result_df = (
  flights_df
  .groupby(['origin_airport'])
  .apply(lambda x: pd.Series({
      'nb_flights': len(x['flight_number'])
  }))
  .reset_index()
)
result_df.head()

origin_airport  nb_flights
0            BOS      105276
1            JFK       91933
2            LAX      192584
3            SFO      145951

### Delayed flights for biggest airlines

Among the biggest airlines, where we define the airline size as the number of yearly flights, we want to know which airline have less delays compared to others

In [17]:
# flight delays
result_df = (
  flights_df
  # Compute airline size and delay statistics
  .groupby(['airline'])
  .apply(lambda x: pd.Series({
      'nb_flights': len(x['flight_number']),
      'perc_delayed': round(np.average(x['is_delayed']), 4)
  }))
  .reset_index()
  # Keep biggest airlines
  .assign(rank_size=lambda x: x['nb_flights'].rank(method='first', ascending=False))
  [lambda x: x['rank_size'] <= 10]
  .drop('rank_size', axis=1)
  # Display airline name
  .merge(airlines_df, left_on='airline', right_on='iata_code')
  .reset_index(drop=True)
)
result_df.head()

airline_x  nb_flights  perc_delayed iata_code               airline_y
0        AA     77024.0        0.1535        AA  American Airlines Inc.
1        AS     14655.0        0.1252        AS    Alaska Airlines Inc.
2        B6     85920.0        0.2080        B6         JetBlue Airways
3        DL     73821.0        0.2016        DL    Delta Air Lines Inc.
4        NK      6475.0        0.3158        NK        Spirit Air Lines

### Temporal evolution of delays by airport

We want to obtain the monthly evolution of delays by origin airport

In [18]:
# monthly evolution of delays by origin airport
result_df = (
    flights_df
    # Create column specific full month name
    .assign(
        month=lambda x: x['scheduled_departure'].dt.strftime('%B'),
        month_num=lambda x: x['scheduled_departure'].dt.strftime('%m').astype(int)
    )
    # Compute aggregation
    .groupby(['origin_airport', 'month', 'month_num'])
    .apply(lambda x: pd.Series({
        'perc_delayed': round(np.average(x['is_delayed']), 4)
    }))
    .reset_index()
    # Reshape data frame
    .pivot_table(
        columns='origin_airport',
        values='perc_delayed',
        index=['month', 'month_num'],
        aggfunc=np.sum
    )
    .rename_axis(None, axis=1)
    .reset_index()
    # Order by month
    .sort_values(['month_num'])
    .drop(['month_num'], axis=1)
    .reset_index(drop=True)
)
result_df.head()

month     BOS     JFK     LAX     SFO
0   January  0.1902  0.2257  0.1738  0.2001
1  February  0.3248  0.3174  0.1978  0.2222
2     March  0.1984  0.2736  0.2246  0.1770
3     April  0.1553  0.2020  0.1855  0.1756
4       May  0.1552  0.1552  0.1990  0.2297

### Month over month percent change in flights

We want to get the month over month percent change in flights

In [19]:
# month over month percent change in flights
result_df = (
    flights_df
    # Create column specific full month name
    .assign(
        month=lambda x: x['scheduled_departure'].dt.strftime('%B'),
        month_num=lambda x: x['scheduled_departure'].dt.strftime('%m').astype(int)
    )
    # Compute aggregation
    .groupby(['origin_airport', 'month', 'month_num'])
    .apply(lambda x: pd.Series({
        'nb_flights': len(x['flight_number'])
    }))
    .reset_index()
    .sort_values(['origin_airport', 'month_num'])
    .reset_index(drop=True)
    # Get lagged values
    .assign(
        nb_flights_before=lambda x:
            x.groupby(['origin_airport'])['nb_flights'].shift(1),
        perc_chance=lambda x:
            round((x['nb_flights'] - x['nb_flights_before']) / x['nb_flights_before'], 2)
    )
    # Last touches
    .drop(['month_num'], axis=1)
)
result_df.head()

origin_airport     month  nb_flights  nb_flights_before  perc_chance
0            BOS   January        8201                NaN          NaN
1            BOS  February        7251             8201.0        -0.12
2            BOS     March        9744             7251.0         0.34
3            BOS     April        9940             9744.0         0.02
4            BOS       May       10172             9940.0         0.02

### Most popular routes by origin airport

We are interested in knowing the top 3 routes per origin airport. We define top routes from a given airport as being routes with the highest number of flights departing from that airport.

In [20]:
# top 3 routes per origin airport
result_df = (
    flights_df
    # Compute volume per route
    .groupby(['origin_airport', 'destination_airport'])
    .apply(lambda x: pd.Series({
        'nb_flights': len(x['flight_number'])
    }))
    .reset_index()
    # Compute rank within each origin airport and keep top ones
    .assign(
        rank_popular=lambda x:
            x.groupby(['origin_airport'])['nb_flights'].rank(method='first', ascending=False)
    )
    [lambda x: x['rank_popular'] <= 3]
    .reset_index(drop=True)
    .drop(['nb_flights'], axis=1)
    .assign(rank_popular=lambda x: 'top_' + x['rank_popular'].astype(int).map(str))
    # Reformat results
    .pivot_table(
        columns='rank_popular',
        values='destination_airport',
        index='origin_airport',
        aggfunc=np.sum
    )
    .rename_axis(None, axis=1)
    .reset_index()
)
result_df.head()

origin_airport top_1 top_2 top_3
0            BOS   DCA   LGA   ORD
1            JFK   LAX   SFO   MCO
2            LAX   SFO   JFK   LAS
3            SFO   LAX   JFK   LAS